### ISSUE TITLE
[v5] Optimize mailbox

### ISSUE THREAD

**Issue by davidkpiano** [statelyai/xstate#2870]
I wonder if we should implement a better data structure for the mailbox, shifting arrays is not the best for performance

_Originally posted by @Andarist in https://github.com/statelyai/xstate/pull/2869#discussion_r767138263_

**Comment by tivac** [comment]
You could track your index in the array, then set `.length = 0` when you're done iterating

**Comment by huan** [comment]
It's great to see that we will have a mailbox wrapper in XState v5!

I'm trying to figure out how to make the async events processing follows the actor pattern (one message at a time) with XState v4 this week, and I have a Mailbox prototype at https://github.com/wechaty/bot5-assistant/blob/actor-pattern/src/mailbox/mailbox.ts

Basically, I'm using three parallel states in the mailbox:

1. `router` to classify the events as `incoming` or `outgoing`
    1. if it's `incoming`, then enqueue it for dispatching
    2. if it's `outgoing`, then `respond` it to the current processing event immediately.
2. `child` to save the `idle` and `busy` state for the child (which is the actual working actor)
3. `message` to check & dispatch the incoming events and only deliver the event to the child when the child is idle (async).

Would love to see if I can contribute any to the v5 mailbox code, and I'm looking forward to using the official v5 mailbox as soon as possible because I'm currently on my way to building a chatbot messaging system based on the XState & actor model.

Thank you very much.

**Comment by Andarist** [comment]
@huan note that the Scheduler used in v4 works sort of like a mailbox for most cases so we still process one message at a time. This is mostly an implementation change in v5 - not a behavior change.

**Comment by huan** [comment]
@Andarist thanks for noticing me that!

Today I have written [a unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L88-L102) to prove that the Mailbox pattern is missing in the XState default machine settings:

1. [#1 unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L60) shows that the default XState machine will lose events when it's in async mode.
2. [#2 unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L108) is using a helper wrapper named `Mailbox.address`, which can queue the events before the actual working actor back to the `idle` state.

This pattern is very useful in my use case because I need to deal with chatbot messages with concurrency in async mode.

Feedbacks and suggestions are welcome.

**Comment by davidkpiano** [comment]
> @Andarist thanks for noticing me that!
> 
> Today I have written [a unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L88-L102) to prove that the Mailbox pattern is missing in the XState default machine settings:
> 
> 1. [#1 unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L60) shows that the default XState machine will lose events when it's in async mode.
> 2. [#2 unit test](https://github.com/wechaty/bot5-assistant/blob/507f3a85c7d958b4a4100188f7cca521e374eb84/src/mailbox/integration.spec.ts#L108) is using a helper wrapper named `Mailbox.address`, which can queue the events before the actual working actor back to the `idle` state.
> 
> This pattern is very useful in my use case because I need to deal with chatbot messages with concurrency in async mode.
> 
> Feedbacks and suggestions are welcome.

Hey @huan, thanks for this, but it's a bit confusing. What specifically are the features you would like XState to support?

**Comment by Andarist** [comment]
Hm, correct me if I'm wrong. The expectation here is what actually happens but you would have expected it to behave differently? If I decode this correctly you are receiving only a single `DONG` event back because one a single `DING` event reaches the `dingMachine` when its state is idle. Other DINGs are delivered to it when it's in its busy state and because of that those messages are being dropped.

From our perspective the mailbox is merely ensuring that we process one event at a time - it's not about queuing messages that were unhandled when being in a particular state. It feels to me like you'd like that to see that latter behavior here and that would be close to the so-called selective receive known from Erlang. 

Would my understanding of this problem be correct?

**Comment by huan** [comment]
Hi @Andarist ,

Thank you very much for letting me know the "selective receive" of Erlang. After I googled it and spend a few hours reading the related links, I created a new Coffee Machine to explain my use case.

Disclaimer: I'm new to both the XState and Actor, have very limited experience with those concepts, so I'm just following my own intuition.

## CoffeeMaker Machine

A naive state machine is a mailbox actor with `capacity=0`, which means it will face the `Dead Letter` problem when new messages come but it has not finished processing the previous one.

Assume we are a coffee maker, and it need 4 three steps to make a coffee:

1. received a `MAKE_ME_COFFEE` event from a customer (sync)
1. get a cup (Async: cost some time)
1. fill coffee into the cup (Async: cost some time)
1. send a `COFFEE` event to the customer (sync)

The coffee maker can only make one cup of coffee at a time, which means that we can not process the `MAKE_ME_COFFEE` event until we have finished the previous one.

The state machine of the coffee maker is:

[![coffee maker machine](https://stately.ai/registry/machines/5bd10d92-3d39-49b5-ad0d-13a8a0a43891.png)](https://stately.ai/viz/5bd10d92-3d39-49b5-ad0d-13a8a0a43891)

If there's a new customer who comes in, and he/she wants coffee, we can get a cup then fill coffee into the cup then deliver a cup of coffee to our customer. So far so good.

## The Problem

However, when two customers come in, and they talk to us at the same time and each customer wants a cup of coffee. After we received the first event(message/request), we are starting to get a cup and can not listen to another event anymore, which will result in an event (the second one) lost (a.k.a Dead Letter).

## The Solution

I'm trying to implement the **Mailbox** concept of the Actor to solve the above problem.

An actor should read the messages to process from its mailbox. A mailbox is an event proxy that holds messages and deals with the backpressure. When the actor can continue processing the next event, it receives the next event from the mailbox.

## The Actor Mailbox Concept I Learned

Actors have mailboxes.

> In the actor model, we must follow that: "if you send 3 messages to the same actor, it will just execute one at a time."
>
> It’s important to understand that, although multiple actors can run at the same time, an actor will process a given message sequentially. This means that if you send 3 messages to the same actor, it will just execute one at a time. To have these 3 messages being executed concurrently, you need to create 3 actors and send one message each.
> 
> Messages are sent asynchronously to an actor, that needs to store them somewhere while it’s processing another message. The mailbox is the place where these messages are stored.
> 
> &mdash; [The actor model in 10 minutes](https://www.brianstorti.com/the-actor-model/)


> an actor is started it will keep running, processing messages from its inbox and won’t stop unless you stop it. It will maintain its state throughout and only the actor has access to this state. This is unlike your traditional asynchronous programming, such as using Future in Scala or promises in javascript where once the call has finished its state between calls is not maintained and the call has finished.
> 
> Once an actor receives a message, it adds the message to its mailbox. You can think of the mailbox as queue where the actor picks up the next message it needs to process. Actors process one message at a time. The actor patterns specification does say that the order of messages received is not guaranteed, but different implementations of the actor pattern do offer choices on mailbox type which do offer options on how messages received are prioritized for processing.
> 
> &mdash; [Introduction to the Actor Model](https://www.softinio.com/post/introduction-to-the-actor-model/)

## A New Mailbox Module

I'm working on a PoC in the past week, comments, and suggestions to the [README](https://github.com/wechaty/bot5-assistant/tree/actor-pattern/src/mailbox) and [code](https://github.com/wechaty/bot5-assistant/blob/actor-pattern/src/mailbox/mailbox.ts) are welcome.

**Comment by huan** [comment]
Hello @davidkpiano ,

Thank you very much for asking me to describe this feature that I want to implement with XState!

I'm keeping studying the great XState framework and trying to implement production actors for my Wechaty chatbot on WeChat, I have described what I want to solve in the above reply to @Andarist , I hope it can help me make my work a little bit clearer.

> What specifically are the features you would like XState to support?

After I have written more code with XState for building Mailbox (with lots of unit tests to help me understand some tricky parts, I'll create another issue later), I think the XState itself is good and my feature request is another problem.

Here's what I think:

1. XState is a state machine management framework, and it did great
2. Actor is a model for processing events both async and distributed
3. Actor and XState have overlap but they are not the same

![xstaet-mailbox](https://user-images.githubusercontent.com/1361891/147311293-8be6f125-16f9-46fd-b3ec-7759e155f8ad.png)

I started learning the Actor concept from the XState docs (thank you so much for the great docs), I really love it and decide to redesign my bot management code to follow the Actor model. I think a Mailbox wrapper is needed to build a standard actor that supports both async and distributed, based on XState.

I'd like to contribute the [Mailbox](https://github.com/wechaty/bot5-assistant/tree/actor-pattern/src/mailbox) I'm working on as an NPM module or PR to the XState core if it can be proved to be useful.

**Comment by huan** [comment]
Update: I have drawn a diagram ([slides link](https://docs.google.com/presentation/d/13oUOIEnzdLWO6KZWztD_pMuu22AQ3SIMjk2wp8f-f18/edit#slide=id.g109fc535984_0_202)) for implementing the "selective receive" pattern with the code updated at [here](https://github.com/wechaty/bot5-assistant/blob/actor-pattern/src/mailbox/mailbox.ts)

![image](https://user-images.githubusercontent.com/1361891/147617671-b536bd55-daf6-4381-990c-aa43b57844cf.png)

Any comments will be greatly appreciated.

**Comment by Andarist** [comment]
I think you are slightly misinterpreting the mailbox and actors. There might be implementations out there that might handle this slightly differently but if we focus on the concepts there is nothing wrong with how XState implements this. It's just that XState works differently from your expectations.

The main difference is that XState **doesn't block** "the thread". When XState receives a message it takes it for processing (and it blocks the message handler from taking anything new from the mailbox, but it doesn't block the mailbox from receiving any new messages in the meantime), it tries to select a transition, if any transition matches then it "executes" it synchronously (which produces a new state, might call some entry/exit/transition actions and stuff like that) and immediately unblocks the "handler" so it can pick up any new messages that are pending in the mailbox.

So if you start any async job within a transition it doesn't block the machine itself from processing messages. This is also compliant with the actor principle that says that processing messages should be quick and shouldn't block the thread for long as otherwise, your mailbox could grow indefinitely. It's important to free processing as quickly as possible.

I don't think there is any step of the overall XState algorithm that would disqualify it from being compliant with basic actor principles.

That being said - your use case is also valid. Both processing algorithms have their place in the actor system. I'm not sure which one is the more common need - I didn't have any problems with what you are asking for. I see the potential of simplifying these kinds of use cases in the XState's core (perhaps by allowing you to configure mailbox type per machine which is allowed by systems like Akka). Because currently you need to implement a lot of manual heavy-lifting to achieve the same goal and this is a low-level concern that could be taken away from your machines and as a result, they could become easier to grok and could focus on what's more important (like the business logic).

**Comment by huan** [comment]
Hi @Andarist ,

Thank you very much for your feedback, it's very valuable. I was keeping working on my mailbox codes for the past 3 weeks, and I will write my thought down and share it with you after they get stabilized for my use case.

**Comment by davidkpiano** [comment]
This has already been done in the `next` branch

**Comment by huan** [comment]
Update:

My mailbox idea has been published as NPM [mailbox](https://www.npmjs.com/package/mailbox) and its GitHub repo at [here](https://github.com/huan/mailbox).

Mailbox is built on top of the XState machine, by adding a message queue to the XState machine and letting the machine decide when to process the next message.

Would love to hear from your guys, any comments, suggestions, and feedbacks are welcome.
